### PR TITLE
Blogpost/team red rat

### PR DESCRIPTION
**Pull request by ubep**

### CODE REVIEW

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+title: "How we test our API"
+date: "2016-09-28 11:11:11"
+image: blog-header/heartbeat.jpg
+categories: tech-stories
+authors: ["Ulf B."]
+---
+
+A few years ago, at the time when the development of the ePages REST API started,
+ePages 6 was just a huge monolith written in Perl and already hard to handle.
+Since the monolith should not be inflated even more and in order to be able to use state-of-the-art frameworks,
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+the REST API was started as a separate service implemented in Java.
+
+This service communicates with the monolith via REST and SOAP in order to operate on the data we won't to expose to the public.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+All the API related business logic like right management and rate limits are implemented in Java such that the API's we had to implement in Perl could
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: geht auch sowas wie "in order to keep the API's we implemented in Perl quiete simple"

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+become quiete simple.
+
+Of course all the resource and services classes are coverey by unit test on both sides, Perl and Java,
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+but since the functionality of the API is based on two different projects that live in two different repositories
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+which have two different teams responsible for the releases,
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+we wanted to have a simple way to check with one click if everything - from end to end - is working as expected.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+This was the starting point of RAT - our Rest API Test framework.
+
+## RAT
+
+The idea of RAT was to create a framework that allows us to easily verify the state of the REST API on an abritrary ePages installation.
+
+In order to do so, RAT sends requests to the API and validates the response.
+To be able to compare the response with an expected one, we let the test run on shop of the DemoShopType
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: wo ist der unterschied?

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+(The developer installation of ePages includes a DemoShopType which allows to create a demo shop with certain products of different types and a few orders).
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+We wanted the test cases to be able to run in a arbitrary order. The problem is that some of the API calls change a shop (DELETE product),
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+which lead to different results in the following calls. The solution is to create new shops for testing those calls.
+The ePages6 REST API does not offer an endpoint to create new shops so far, so we used the ePages SOAP API to do that.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+## RAT
+
+The idea of RAT was to create a framework that allows us to easily verify the state of the REST API on an abritrary ePages installation.
+
+In order to do so, RAT sends requests to the API and validates the response.
+To be able to compare the response with an expected one, we let the test run on shop of the DemoShopType
+(The developer installation of ePages includes a DemoShopType which allows to create a demo shop with certain products of different types and a few orders).
+We wanted the test cases to be able to run in a arbitrary order. The problem is that some of the API calls change a shop (DELETE product),
+which lead to different results in the following calls. The solution is to create new shops for testing those calls.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+The ePages6 REST API does not offer an endpoint to create new shops so far, so we used the ePages SOAP API to do that.
+
+RAT is based on [REST-assured](http://rest-assured.io/) and [Serenity BDD](http://www.thucydides.info), two tools that take over a lot of work on the way to produce nice test results. In the next paragraphs I will introduce these tools briefly and show how we use them.
+
+### REST-assured
+
+The certain test cases are implemented using REST-assured. It is a framework that is specially designed for testing REST APIs and allows easily to send different kind of requests and validation of their responses.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+
+A simple check if the product resource of our REST API is returning 22 products could look like that:
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+    .and()
+    .body("results", equalTo(22));
+{% endhighlight %}
+
+So REST-assured allows us to do request and validation in one step (*assertThat()* and *and()* are syntactic sugar and could be left out).
+
+In RAT we usually store the result for later use (e.g. extracting the links and follow them with further requests)
+and do the validation in a separate step.
+
+Since baseUri and request headers are for every request more or less the same, we put the request specification in a separate class:
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+.when()
+    .get("products")
+.then()
+    .assertThat()
+    .statusCode(200)
+    .and()
+    .body("results", equalTo(22));
+{% endhighlight %}
+
+So REST-assured allows us to do request and validation in one step (*assertThat()* and *and()* are syntactic sugar and could be left out).
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+                    .and() //
+                    .extract() //
+                    .body() //
+                    .jsonPath();
+{% endhighlight %}
+
+
+#### JSONPath ####
+
+REST-assured comes already with a library for JSONPath (which is basically [XPath](https://en.wikipedia.org/wiki/XPath) for JSON) namely [com.jayway.restassured:json-path](https://mvnrepository.com/artifact/com.jayway.restassured/json-path),
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+that can be used to validate a JSON response and verify certain attributes.
+
+Unfortunately this library does not implement all the JSONPath expressions from [Stefan Gössner's JSONPath specification](http://goessner.net/articles/JsonPath/) - that's why we use [com.jayway.jsonpath:json-path](https://mvnrepository.com/artifact/com.jayway.jsonpath/json-path) instead.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+To use this library we cannot simply do *... .extract().body().jsonPath();* anymore. Instead we have to get the response as a string first and then parse it in a second step:
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: "parse it in:" das "in" am Ende ist komisch, vielleicht besser "and to parse it in a second step" ?

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+// or
+List<String> productNames = products.read("items[*].name");
+{% endhighlight %}
+
+
+
+### Serenity BDD
+
+Similar to [Cucumber](https://cucumber.io/) in the Ruby world,
+Serentiy allows you to describe and structure the test cases in way that acts like a complete specification that describes each use case step by step.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+        steps.when_uploading_a_product_image(IMAGE_NAME);
+        steps.then_the_product_slideshow_contains_image(IMAGE_NAME);
+        steps.when_the_product_image_is_changed_with_slideshow_image(IMAGE_NAME);
+        steps.then_the_product_image_is_changed_accordingly();
+    }
+
+}
+{% endhighlight %}
+
+The certain test steps are in a separate classes. We have a class *ShopSteps* which contains tests that are shop-specific and needed in all tests. Then we have e.g. *ProductSteps* inherting from ShopSteps for steps that are used in product-related tests.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: vielleicht besser "The implementation of the test steps can be found in different classes. " ?

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+
+This example shows the *ProductImageSteps* (which for their part inherit from *ProductSteps*) which only contains those steps that are relevant for product image-related tests:
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    @Step("Then the product slideshow contains image '{0}'")
+    public void then_the_product_slideshow_contains_image(String imageName) {
+        // ... get slideshow using REST-assured and check if the imageName appears in the result ...
+    }
+}
+{% endhighlight %}
+
+Serenity offers the *@Step* annotation to define steps. The string in the argument will appear later in the test report. You can use the variable *{n}* which will be replaced with the n-th argument the method is called with.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+
+#### HTML test report
+
+After running all the test, Serenity generates a HTML test report. On its homepage, the report offers a nice pie chart showing the test results of each feature category. The following image shows how this diagram currently looks like in our project.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 25]
```diff
@@ -0,0 +1,191 @@
...
+
+{% image blog/blog-rat-serenity-results.png %}
+
+We try to work test-driven, therefore there are a few test cases in a *pending* state (To do that, Serentity offers a *@Pending* annotation).
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 26]
```diff
@@ -0,0 +1,191 @@
...
+These tests are already written but will be skipped since the features they should test are not yet implemented. As soon as everything is in place, they can be activated.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 27]
```diff
@@ -0,0 +1,191 @@
...
+
+You can click on a feature to see what kind of test cases are there. Each test case can consist of multiple steps which can be displayed in detail:
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 28]
```diff
@@ -0,0 +1,191 @@
...
+
+{% image blog/blog-rat-serenity-results-slideshow.png %}
+
+In this test case we wanted to check if an image that was uploaded for a product is appearing in the product slideshow. The first step gives us a DemoShop (which can be newly created or was already used in other read-only test cases). Then we take the first product from the product collection resource (*GET /products*) and upload the image. Finally we retrieve the product slideshow (*GET /products/{productId}/slideshow*) and check if the uploaded image is in there.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md (hunk 8 above)
Comment (roseritter) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: das "a"..also: "we let the test run on **a** shop of the DemoShopType"

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 29]
```diff
@@ -0,0 +1,191 @@
...
+The idea of RAT was to create a framework that allows us to easily verify the state of the REST API on an abritrary ePages installation.
+
+In order to do so, RAT sends requests to the API and validates the response.
+To be able to compare the response with an expected one, we let the test run on shop of the DemoShopType
+(The developer installation of ePages includes a DemoShopType which allows to create a demo shop with certain products of different types and a few orders).
+We wanted the test cases to be able to run in a arbitrary order. The problem is that some of the API calls change a shop (DELETE product),
+which lead to different results in the following calls. The solution is to create new shops for testing those calls.
+The ePages6 REST API does not offer an endpoint to create new shops so far, so we used the ePages SOAP API to do that.
+
+RAT is based on [REST-assured](http://rest-assured.io/) and [Serenity BDD](http://www.thucydides.info), two tools that take over a lot of work on the way to produce nice test results. In the next paragraphs I will introduce these tools briefly and show how we use them.
```
Comment (roseritter) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ahh..Das hatte ich falsch verstanden. Dann würde man sagen: "RAT is based on REST-assured and Serenity BDD, two tools that ease a lot of our workload and achieve good test results. In the following paragraphs, I will introduce the mentioned tools briefly and show how we use them."

File: _posts/2016/2016-09-28-rest-api-test-framework.md (hunk 19 above)
Comment (roseritter) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ja, das geht auch!

File: _posts/2016/2016-09-28-rest-api-test-framework.md (hunk 21 above)
Comment (roseritter) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ja, das ist gut!

File: _posts/2016/2016-09-28-rest-api-test-framework.md (hunk 3 above)
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: All API related business logics such as right management and rate limits are implemented in Java in order to make sure that the Perl implementation of the APIs is simple.

Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md (hunk 29 above)
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 30]
```diff
@@ -0,0 +1,191 @@
...
+
+{% image blog/blog-rat-serenity-results-slideshow.png %}
+
+In this test case we wanted to check if an image that was uploaded for a product is appearing in the product slideshow. The first step gives us a DemoShop (which can be newly created or was already used in other read-only test cases). Then we take the first product from the product collection resource (*GET /products*) and upload the image. Finally we retrieve the product slideshow (*GET /products/{productId}/slideshow*) and check if the uploaded image is in there.
+
+
+
+## Summary
+
+RAT allows us to check the complete state of the REST API on an abritrary ePages installation just at the push of a button. We use it to partly automate our QA process and run it as last step of a developer installation via Jenkins to make sure everything went fine. RAT takes over repetitive tasks and guarantees a certain level of quality - we can spend the saved time on improving our test infrastructure and on the implementation of further test cases.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

File: _posts/2016/2016-09-28-rest-api-test-framework.md [hunk 31]
```diff
@@ -0,0 +1,191 @@
...
+
+Currently we discuss how to include RAT in the Jenkins job that checks our GitHub pull requests for new features in the Perl or Java repository. The challenge here is to choose the RAT branch that corresponds with the changes in the branch from the pull request.
```
Comment (ubep) [inline _posts/2016/2016-09-28-rest-api-test-framework.md]: ok

**Pull request merged**